src/Microsoft.Extensions.Hosting/Internal/ApplicationLifetime.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Hosting.Internal
+{
+    /// <summary>
+    /// Allows consumers to perform cleanup during a graceful shutdown.
+    /// </summary>
+    public class ApplicationLifetime : IApplicationLifetime
```

Comment:
@Tratcher I think this should be implemented on the IHost itself.

Reply:
Let's punt that discussion